File: 2022/02/17/remove_duplicates.py
"""
https://leetcode.com/problems/remove-duplicates-from-sorted-array/
Author: Andrew Jarombek
Date: 2/17/2022
"""

from typing import List


def remove_duplicates(nums: List[int]) -> int:
    prev = -101
    end = 0

    for index, num in enumerate(nums):
        if num != prev:
            prev = num
            nums[end] = num
            end += 1

    return end

File: 2022/02/27/cpp/removeDuplicates.cpp
// https://leetcode.com/problems/remove-duplicates-from-sorted-array/
// Author: Andrew Jarombek
// Date: 2/27/2022

#include <vector>
#include <cassert>

using namespace std;

class Solution {
public:
    static int removeDuplicates(vector<int>& nums) {
        int slow = 0;
        int fast = 1;

        while (fast < nums.size()) {
            if (nums[fast] != nums[slow]) {
                nums[++slow] = nums[fast];
            }

            fast++;
        }

        return slow + 1;
    }
};

int main() {
    vector<int> v {1, 1, 2, 3};
    auto result = Solution::removeDuplicates(v);
    assert(result == 3);
    assert(v[0] == 1);
    assert(v[1] == 2);
    assert(v[2] == 3);
}
